### ISSUE TITLE
Add options to only sync visible documents to the server

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode-languageserver-node#848]
This is slightly related to https://github.com/microsoft/vscode/issues/15178. Right now there's no way for an editor to know when a document is either "open" or "visible" in a way that matches the users expectation:

> Open in this sense means it is managed by the client. It doesn’t necessarily mean that its content is presented in an editor.

There are many situations where a server would like to know what files a user has open (or visible). For example:

- The user wants to ignore lints/errors in certain files - for example files generated by tools
- When the user opens the file in the editor, they want to see these errors/lints (eg. to debug something going wrong in them)
- When the user closes the file, they expect the errors/lints to disappear (immediately, not after 3 minutes)

Even though VS Code doesn't have an API for "open editors", it does have an API for "visible editors". Even if VS Code plans to never support "open editors" it would be very useful for LSP to provide "visible editors". On the other hand, if VS Code *does* add support for "open editors", then that could be provided instead/as well. In either case, it seems like LSP needs some changes to support one or the other (or both).

(as a workaround, I'm considering implementing custom messages in the VS Code extension + LSP server to pass the "visible editors" over to the server.. it feels really hacky though).

**Comment by ljw1004** [comment]
> For the more general fix. When VS Code has a proper tab model I will add support to the LSP libs to configure if open should be send for invisible files or not. But this will be an implementation detail in the extension / libs.

@dbaeumer could you expand on this? Will the option also control whether didClose is sent for invisible files? And will it also control whether (say) documentSymbols is sent for invisible files?

(I'm asking because I'm implementing a hacky temporary middleware which suppresses didOpen+didClose in the case that it's an invisible tab, and I'd like my middleware to have the same behavior as what you eventually envisage for your LSP option.)

**Comment by dbaeumer** [comment]
@ljw1004 The idea is to not send and open/change/close events for documents that are not visible in an editor. However since LSP allows sending language requests for not opened documents those will still be allowed on these documents. Sending a language request might need to trigger and open/change notification in case the document is dirty on the client.


Instead of putting this into your own middleware would you consider providing a PR for this?

**Comment by mhuisi** [comment]
What is the current state of this issue? Is there any workaround that works and reliably filters exactly the invisible documents?

Not spamming the language server with `didOpen` notifications is important for us in the language server for the [Lean 4 programming language and theorem prover](https://lean-lang.org/) since processing documents after `didOpen` can be potentially very expensive (e.g. Lean needs to execute meta-programs, arbitrary proof automation, dependent type checking, higher order unification and typeclass search so that we can provide diagnostics and respond to requests for the file). We definitely do not want to initiate this process whenever someone hovers over an identifier while holding `Ctrl`.

A list of things that we have tried in the past to work around this issue:
- Filter using `window.visibleTextEditors` in the `didOpen` middleware. This does not work because `window.visibleTextEditors` is not up-to-date when the `didOpen` middleware is called.
- Filter all `didOpen` notifications in the `didOpen` middleware and add our own `window.onDidChangeActiveTextEditor` and `window.onDidChangeVisibleTextEditors` handlers. This does not work because it will occasionally cause a race condition where the language client library attempts to emit requests to the file before our own handlers fired and opened it. See https://github.com/leanprover/vscode-lean4/pull/460 for details.
- Filter using `window.tabGroups.all` in the `didOpen` middleware. This often works, but there are some cases where the `didOpen` middleware fires before `window.tabGroups` is updated, e.g. in the sequence of steps in https://github.com/leanprover/vscode-lean4/issues/484 where an `untitled` plaintext file is saved as a Lean document.

I don't see any other VS Code API to filter these requests and I'm wary of messing with the control flow of the language client library since it may introduce undesired race conditions. What is the recommended way to mitigate this issue?

**Comment by dbaeumer** [comment]
@mhuisi what you can do right now is the following:

- don't do anything on a didOpen in the server
- wait for a concrete request to do the expensive work
- switch from push diagnostics to pull diagnostics to implement diagnostics in that model. The pull diagnostic model will only pull for visible documents.

**Comment by mhuisi** [comment]
> @mhuisi what you can do right now is the following:
> 
>     * don't do anything on a didOpen in the server
> 
>     * wait for a concrete request to do the expensive work
> 
>     * switch from push diagnostics to pull diagnostics to implement diagnostics in that model. The pull diagnostic model will only pull for visible documents.

We can't switch to pull diagnostics because AFAIK [VS Code still does not support result streaming / partial results](https://github.com/microsoft/vscode/issues/105870). As mentioned earlier, processing Lean files is expensive and can take a while, so we can't have users wait for diagnostics for (up to) minutes until the diagnostic pull request can be responded to.

**Comment by dbaeumer** [comment]
So you have a first error after some seconds and the last once after minutes for one file?

**Comment by mhuisi** [comment]
> So you have a first error after some seconds and the last once after minutes for one file?

Yes, because semantic analysis is (necessarily) very expensive in languages like Lean. For a concrete example, proof scripts may use SMT decision procedures to prove lemmas, which in turn can be used by the type checker to ensure that a program satisfies some formal specification. We also implement an additional [`$/lean/fileProgress` push notification](https://github.com/leanprover/vscode-lean4/blob/master/vscode-lean4/manual/manual.md#file-processing) that displays an orange text editor decoration in the overview ruler / gutter so that users aren't lost about which parts of the file have already been processed.

There are [some other extensions](https://github.com/leanprover/vscode-lean4/blob/master/vscode-lean4/manual/manual.md#infoview) we make to LSP to suit interactive theorem proving needs as well, but those are not really relevant here.

Note that the streaming support issue also has a comment from one of the maintainers of coq-lsp, which is another interactive theorem prover operating under similar constraints as Lean.

**Comment by dbaeumer** [comment]
You can split the diagnostics publishing into two stages even with the new pull model. You can register two pull providers with different identifiers which give you two independent channels to report diagnostics (see4 https://github.com/microsoft/vscode-languageserver-node/blob/release/sorry-butterfly-pink/protocol/src/common/protocol.diagnostic.ts#L64)

**Comment by mhuisi** [comment]
> You can split the diagnostics publishing into two stages even with the new pull model. You can register two pull providers with different identifiers which give you two independent channels to report diagnostics (see4 [`release`/sorry-butterfly-pink/protocol/src/common/protocol.diagnostic.ts#L64](https://github.com/microsoft/vscode-languageserver-node/blob/release/sorry-butterfly-pink/protocol/src/common/protocol.diagnostic.ts?rgh-link-date=2024-06-25T07%3A06%3A10Z#L64))

I would really prefer not to keep piling on additional hacks. It's also not clear to me that this will give us the kind of fine-grained diagnostic reporting that users expect from tools like Lean.

Can you please clarify why e.g. expecting the `tabGroups` API to accurately reflect the state of the tabs that the user sees when the middleware fires is unreasonable?

**Comment by dbaeumer** [comment]
> I would really prefer not to keep piling on additional hacks. 

This is actually not a hack. Other language servers do the same to separate syntax errors from more expensive semantic errors.

The eventing is tricky and I try to stay away form defining in LSP in which order they have to arrive when VS Code events (emitted sync) are combined with LSP middleware call chains (which are async). This is why for example LSP emits its own open sent events instead of users of the libs relying on the `didOpen` event from VS Code.

Even if it is reflected correctly (there might be a bug / missing event when the input of a editor changes) I still would stay away from specify a event order.

**Comment by dbaeumer** [comment]
The pull diagnostics abstracts from this by listening to open events and tab change events and pulls for the document independent on how it got to live (minus some bugs since as said the whole thing is tricky :-)).

**Comment by mhuisi** [comment]
> > I would really prefer not to keep piling on additional hacks.
> 
> This is actually not a hack. Other language servers do the same to separate syntax errors from more expensive semantic errors.
> 
> The eventing is tricky and I try to stay away form defining in LSP in which order they have to arrive when VS Code events (emitted sync) are combined with LSP middleware call chains (which are async). This is why for example LSP emits its own open sent events instead of users of the libs relying on the `didOpen` event from VS Code.
> 
> Even if it is reflected correctly (there might be a bug / missing event when the input of a editor changes) I still would stay away from specify a event order.

Ah. Unfortunately, all errors in Lean are potentially expensive semantic errors, so there is no simple two-stage separation here that is independent of the progress the server makes on processing the file. Due to Lean's meta-programming features (which are especially helpful for interactive theorem proving!), we cannot proceed with parsing the next command until we have completed the semantic analysis for the previous command, since it may modify the environment of the parser. Hence, given these constraints, I initially understood your suggestion as "use the first pull provider for the first half of the file and the second pull provider for the second half of the file", which I would still refer to as a hack :-)

> So it is even not something I can fix in LSP without re-interpreting event orders.

Should we report this issue to VS Code, or do you think it is hopeless?

> The pull diagnostics abstracts from this by listening to open events and tab change events and pulls for the document independent on how it got to live (minus some bugs since as said the whole thing is tricky :-)).

In any case, even if this discussion won't resolve our issues with LSP (at least until VS Code gets streaming support for these requests), thank you for your helpful responses.

**Comment by DanTup** [comment]
> Should we report this issue to VS Code, or do you think it is hopeless?

If you specifically mean it sending open/close when you hover, I think it's hopeless :) I tried that here:

https://github.com/microsoft/vscode/issues/109908

It was closed as a dupe of https://github.com/microsoft/vscode/issues/15178 which has since been implemented.

It affected TypeScript too:

https://github.com/microsoft/vscode/issues/84875

IMO this is a hack. Language servers may want to know about open documents because a) this changes the owner of the document from the disk to the editor and b) they might want to prioritise or cache data related to open files for performance. VS Code sending an open immediately followed by a close when it's just reading the contents from disk because it might show a preview in the tooltip doesn't feel like it needs to trigger any of this, it feels like an implementation detail of VS Code leaking into language servers (either they specifically write code to detect and ignore this, they potentially suffer worse performance because they start work they didn't need to, or they potentially don't send notifications for push-based features until some other request for the file shows up) 😔

**Comment by ljw1004** [comment]
I work on the Hack LSP server. It also takes time (1s-5mins depending on complexity) to fully recompute diagnostics. My solution:
1. Maintain a list of open files
2. For each didOpen/didChange, mark the file as "dirty"
3. When the LSP is idle (i.e. has no other messages to process), then recompute diagnostics for one of the file's that's dirty.

This solves the transient didOpen/didClose problem nicely. It also offered a HUGE performance benefit for our single-threaded LSP server, because when a user opened a fresh document and did (say) hover/gotodef then the LSP server would be computing latency-sensitive user actions rather than non-critical squiggles.

When I talk about "performance" I refer to what percentage of hover/gotodef/autocomplete requests get a useful response within 200ms (which is my team's criterion for "adequately fast"). Once I ceased doing work on didOpen, and changed to this lazy model, then we went from ~90% of requests getting responses in under 200ms up to ~97%.

I spent a lot of time experimenting with pull diagnostics and came to the conclusion that they weren't as good a solution. There are two chief reasons: (1) VSCode sends pull diagnostic requests more *eagerly* than my solution, hence doesn't give as much benefit, (2) more importantly there's no race-free way to combine pull diagnostics with the push diagnostics inherent in a up-to-5-minute error recomputation, although this might be under-specification rather than an inherent blocker. Here are my raw notes:

**Normal VSCode didChange behavior.** When you change an unmodified file on disk, VSCode updates the editor contents and sends didChange; when you change a modified file on disk, VSCode doesn't update the editor contents and doesn't send didChange. Thus for instance if you rebase at the CLI, then VSCode will send didChange for all files modified by the rebase.

Also, imagine a file on disk has errors in it, but the modified buffer doesn't, and you close the modified buffer. The user will naturally expect the diagnostic to be restored upon file-close. Well, this is helped by some (undocumented) VSCode behavior that (1) immediately prior to didClose, it first synthesizes a didChange event which restores the editor content to what is on disk, apparently via line-diffing. Then it sends a pull diagnostic request for the file. Then (without waiting for a response) it sends didClose.

**InterFileDependencies.** This is a flag you specify in your server capabilities. It means that a change in one file will give rise to pull diagnostics requests not just for that file, but for all open files. This is true both in the rebase scenario and regular typing. Therefore, rebases will cause exactly the same set of files to get pulled-diagnostic as got didChange.

VSCode has weird behavior (bug?) where it doesn't respect already-open-upon-project-start files for its interFileDependencies. Specifically: if you start with editor tabs open for A and B, and change A, then it will pull diagnostics for A but not B. But if you subsequently switch to tab B and back again to A and change A again, then it will now pull for both A and B.

**Sequencing.** VSCode sends pull-diagnostics request immediately after sending didChange, but it seems to wait until the previous pull-diagnostic has gotten a response before sending another request. So (1) if there are multiple editor tabs open then it only requests diagnostics one at a time, (2) if you type really fast then it will send didChange for each keystroke but will consolidate and not have a backlog of pull diagnostic requests.

**Disk file changes? Races?** If you change a file on disk that's not open in the editor, VSCode won't pull any diagnostics (even if you have interFileDependencies turned on). I suppose you must have to use other mechanisms. I don't know what they'd be, i.e. I don't know how my didChangeWatchedFiles handler should trigger the whole pull-diagnostics machinery. Or alternatively, if my LSP responds to this didChangeWatchedFiles event by pushing diagnostics, how will that race with any pull diagnostics that VSCode sends for the file?

Another race I don't understand is if we close a file then VSCode sends didChange followed by a pull-diagnostics request followed immediately by didClose. What if the LSP feels free to push diagnostics for the file after didClose, but VSCode acts upon receipt of pull-diagnostics after that?

Failure? I didn't yet investigate how pull diagnostics works with failure, e.g. if a pull-diagnostic request gets an error, then are any previously pushed diagnostics for this file removed? are any previously pulled diagnostics for this file removed?

All in all, I actually don't see any benefit to pull diagnostics. I think I could achieve the same overall effect by having the LSP server store a list of files which have so far received outstanding didChange/didOpen events, and if it's idle then push diagnostics for them as needed. I could similarly make didChangeWatchedFile add open files into the list to refresh them. This way I wouldn't worry about races.

(The only thing I'd lose with this approach is that pull diagnostics tells me which file should be prioritized).

**Comment by mhuisi** [comment]
@ljw1004

Unfortunately, I don't think this approach will work for us, as it doesn't really resolve the issue of `didOpen` requests being emitted for invisible files and subsequently putting pressure on the language server. 
We also have no way of separating "processing for diagnostics" and "processing for all other requests" into two separate phases, and the former is by far the most important language server feature for languages like Lean anyways. Additionally, our language server uses multiple processes (one for every open file and one to manage these processes and handle requests that require global information) to reduce overall latency and because of certain technical reasons (e.g.: we `mmap` data from imports into the process and we need to be resilient against buggy meta-programs and proof automation).

Thanks for your detailed analysis and the word of warning about pull diagnostics.

**Comment by mhuisi** [comment]
@DanTup I was referring to the `tabGroups` event ordering issue, since (from my understanding) the `tabGroups` API was always suggested as the future solution to mitigate these issues in the past.

**Comment by DanTup** [comment]
>  I was referring to the `tabGroups` event ordering issue, since

Oh, you mean so we could use middleware to filter out `didOpen`s for "invisible" documents? That seems valid to me, and is not what I'd already asked for. (I also think if that's solved, it would be nice to support it here rather than in extension-specific middleware). If you open this issue, please post the link here so those of us also interested can add 👍.

**Comment by ljw1004** [comment]
> Unfortunately, I don't think this approach will work for us, as it doesn't really resolve the issue of didOpen requests being emitted for invisible files and subsequently putting pressure on the language server.

@mhuisi The reason it works for us (solves the ephemeral didOpen pressure) is because for those ephemeral cmd+hovers, VSCode sends didOpen followed *immediately* by didClose. Our LSP server first handles `didOpen(file1)` by adding `file1` to the list of open files and setting its dirty bit. The immediate next thing it does is find `didClose(file1)` in stdin and so removes `file1` from the list of open files. Now our LSP server is idle, so it looks for all dirty files in the list of open files, and doesn't find any, hence experiences no pressure. I'd expected a similar approach to work for you?

(There is a theoretical race condition here, that LSP process might discover empty stdin because the VSCode process I guess got switched out before it had a chance to write didClose, but our telemetry never showed that to happen in practice).

By the way, I earlier tried `tabGroup` approach suggested in https://github.com/microsoft/vscode-languageserver-node/issues/683#issuecomment-1236347083 -- it appeared to work in testing, but once I shipped to production it failed bad enough that we had to do an emergency rollback. Here was my attempted implementation:

```
/**
 * VSCode behavior is that when you cmd+hover over a symbol, it sends textDocument/definition
 * request, then quickly fires off didOpen+didClose.
 * This has been a long-standing complaint https://github.com/microsoft/vscode-languageserver-node/issues/683
 * because LSPs often do costly work in response to didOpen.
 *
 * Ultimately the plan is for LSP to configure whether it wants to receive didOpen for these ephemeral
 * didOpens https://github.com/microsoft/vscode-languageserver-node/issues/848#issuecomment-960636388.
 * In the meantime, Typescript uses a workaround https://github.com/microsoft/vscode/commit/fa72810c777c07b8589ca0495d343e057d180977
 * which examines the pending queue of messages and if there's a didOpen followed by a didClose
 * then don't bother handling it. Alas we can't use the same workaround because we don't have visibility
 * into the queue, and at the time our didOpen is executed the didClose hasn't yet been sent.
 * So this middleware implements a different workaround suggested in
 * https://github.com/microsoft/vscode-languageserver-node/issues/683#issuecomment-1236347083
 * Namely: when we get a didOpen, check if it truly corresponds to an open tab,
 * and if not then it must have been one of those ephemeral ones. We'll try to replicate
 * future planned LSP behavior by suppressing didOpen (and didClose) for these cases.
 * What should be the behavior if VSCode wants to send other messages too? like documentSymbol
 * or codeAction? I don't know. With this diff I'll just let them through to the LSP.
 * Hopefully the LSP will respond "sorry, action on non-open document, won't do anything."
 * (I don't want this middleware to have to enumerate every single possible message
 * and report the error itself.)
 * Incidentally, the caller kicks off didOpen and didClose in succession, so if our didOpen middleware
 * has any awaits e.g. to find project root, in it then the didClose could reach the client prior
 * to the didOpen! We've implemented it without awaits.
 */
export class EphemeralOpenMiddleware implements Middleware {
  private suppressedUris = new Set<vscode.Uri>();

  constructor() {
    // without the next lines, `this` will be unbound when the middleware functions are called
    this.didOpen = this.didOpen.bind(this);
    this.didClose = this.didClose.bind(this);
  }

  didOpen(
    document: vscode.TextDocument,
    next: (data: vscode.TextDocument) => Promise<void>,
  ): Promise<void> {
    const documentIsInSomeTab = vscode.window.tabGroups.all.some(group =>
      group.tabs.some(tab => document.uri.fsPath.endsWith(tab.label)),
    );
    if (documentIsInSomeTab) {
      return next(document);
    } else {
      this.suppressedUris.add(document.uri);
      return Promise.resolve(undefined);
    }
  }

  didClose(
    document: vscode.TextDocument,
    next: (data: vscode.TextDocument) => Promise<void>,
  ): Promise<void> {
    if (this.suppressedUris.has(document.uri)) {
      this.suppressedUris.delete(document.uri);
      return Promise.resolve(undefined);
    } else {
      return next(document);
    }
  }
}
```

What we observed is that in production, for reasons I never managed to reproduce mlocally, many of our users encountered a legitimate didOpen (i.e. they truly did open a file) for which `documentIsInSomeTab === false`. This middleware therefore suppressed it and our language server believed the file was never opened, i.e. we never added it to the LS's open-file-list. But when a subsequent hover/gotodef/completion request came in for that file, well, our LS errors when it receives a request on a non-open file. The number of users who reported "hover sometimes just doesn't work" was high, and it showed up clearly in our telemetry. My conclusion was that the semantics of tabGroups isn't precisely enough defined for me to rest critical code upon it!

**Comment by mhuisi** [comment]
@ljw1004 
> The reason it works for us (solves the ephemeral didOpen pressure) is because for those ephemeral cmd+hovers, VSCode sends didOpen followed immediately by didClose. Our LSP server first handles didOpen(file1) by adding file1 to the list of open files and setting its dirty bit. The immediate next thing it does is find didClose(file1) in stdin and so removes file1 from the list of open files. Now our LSP server is idle, so it looks for all dirty files in the list of open files, and doesn't find any, hence experiences no pressure. I'd expected a similar approach to work for you?

Thanks! I had always assumed that there are `didOpen` notifications for invisible documents that are not closely followed by a `didClose`. If that is indeed always the case (though it is clearly a bit of a brittle assumption to make), we could also debounce `didOpen` notifications that are closely followed by a `didClose` notification in our language server without making any additional adjustments to the language server architecture.

> What we observed is that in production, for reasons I never managed to reproduce mlocally, many of our users encountered a legitimate didOpen (i.e. they truly did open a file) for which documentIsInSomeTab === false.

You can find one reproducible example of this in this thread above that was reported by one of our users (saving an untitled file). That's what the `tabGroups` event order discussion was about.

**Comment by dbaeumer** [comment]
@ljw1004 @mhuisi @DanTup great discussion. I would like to add a couple of comments:

- it does make sense to not do expensive work immediatelly on didChange. I did the same for ESLint until I moved to pull diagnostics which basically do that for me on the client side for free.
- there can be `didOpen` events for documents that are not visible in an editor that stay open for a long time. `didOpen` is in hindsight a bad name since the events purpose is to signal content ownership transfer from the server to the client for a specific document. This usually happens when an editor opens but can also happen in other situations. 

Regarding the tab eventing in VS Code. Looking at the event order I would not argue that there is something wrong in VS Code. In the case of saving a untitled file, VS Code generates a new document, however since the editor already exists, it only receives a new input. That results in a changed event on the tab API but on a close/open event on the document. This is different from opening an editor, where the editor first gets created and then the input is created and set.

So if you want to handle this in the middleware you basically need to do the same as the client does for pull diagnostics:

- if you receive a open event check if it is part of the tabs. If yes, forward if no drop
- have a listener on the tab model. If an editor opens or changes check if the document is know on the server. If not, send an open event to the server (you can use getFeature to get the open feature to use the same API: https://github.com/microsoft/vscode-languageserver-node/blob/main/client/src/common/client.ts#L1712)
- if a editor closes or changes check if the document is know on the server and if yes, send a close event.
- if a document closes, check if the document is know on the server. If not, drop the document other send the event.

This is basically what pull diagnostic does.

I haven't implemented that behind a flag since I am not convinced anymore that only syncing visible document to the server is a good approach  (there can be long lived documents that are not visible and relevant for a server).

**Comment by mhuisi** [comment]
> there can be `didOpen` events for documents that are not visible in an editor that stay open for a long time. [...] there can be long lived documents that are not visible and relevant for a server

Do you have an example?

**Comment by DanTup** [comment]
@dbaeumer 

> it does make sense to not do expensive work immediatelly on didChange. I did the same for ESLint until I moved to pull diagnostics which basically do that for me on the client side for free.

As far as I can tell, pull diagnostics means VS Code chooses which files to request diagnostics for, but we want users to see all errors across their entire project. When using push diagnostics, we have to reanalyze in `didChange`.

> `didOpen` is in hindsight a bad name since the events purpose is to signal content ownership transfer from the server to the client for a specific document. 

Given this, why is it normal for a client to send `didOpen` immediately followed by `didClose`? There is no change of ownership here because the close happens immediately. Given the two cancel each other our, the notifications are completely pointless and just add grief.

IMO if VS Code won't fix this, it seems it would be better done here in the LSP client than everyone affected having to write middleware to try and strip this out. (note: I'm not saying the LSP client should necessarily sync only visible files as this issue became, but I do think it could drop these pointless open/closed events that are a VS Code specific quirk and don't make sense over LSP).

**Comment by ljw1004** [comment]
> When using push diagnostics, we have to reanalyze in didChange.

@DanTup I don't think that's true. What we do is that didChange merely sets an "is-dirty" bit in our LS. Then the LS can perform its reanalysis (so clearing the dirty bit) at its leisure, at any time it choses.

**Comment by DanTup** [comment]
> @DanTup I don't think that's true. What we do is that didChange merely sets an "is-dirty" bit in our LS. Then the LS can perform its reanalysis (so clearing the dirty bit) at its leisure, at any time it choses.

Actually we do similar - but in some cases there is some synchronous work we must do (to ensure consistency), and for some configurations we modify our analysis roots based on open files (to support large workspaces where we _don't_ want to analyze everything). We could certainly work around this, but I just don't understand why this is being considered normal. If the request is for transferring ownership to the server, why would it makes sense to send didOpen/didClose immediately just because the user Ctrl+Clicked on a symbol and holding Ctrl happened to trigger a hover and the hover wanted to show a preview of contents?

This feels like a weird quirk/implementation detail of VS Code that doesn't make sense over LSP and servers are having to change their behaviour to try and mitigate the effects of. Why don't we just filter it out here? There is (as far as I can tell) zero benefit in sending those two events over the wire in this case.

**Comment by DanTup** [comment]
That sounds good to me, thank you! :)

Will these releases be tied to the next version of LSP? (it's not clear to me what the release schedule is for LSP, I think it's been a while since the last release and I haven't tried to use a client for the new version yet).

**Comment by mhuisi** [comment]
> I looked into it and here is what I think would make sense to implement to not make things overly complicated:
> 
>     * if a document is opened and it is NOT visible in a editor the open event will be delayed
> 
>     * if the editor opens or any other request / notification is trigger (except a close of the delayed document) the open notification will be forwarded to the server.
> 
> 
> This behavior will go behind a flag in the client options.
> 
> First cut of the code is in `dbaeumer/bottom-eagle-chocolate`
> 
> If you think that this would help for your scenarios I will make next releases so that you can give it a try.

Thanks, sounds good! :-)

**Comment by dbaeumer** [comment]
The new client version `10.0.0-next.9` has the mention support. `LanguageClientOptions#textSynchronization.delayOpenNotifications` is the flag to control the behavior. 

Please note that the way how the client and server exports their API has changed. It is now based on the `exports` property in the package.json to allow further move to new features. See the Readme for details.

**Comment by DanTup** [comment]
@dbaeumer I tested this out and it seems to work as described. If I hold ctrl and hover over a bunch of symbols without clicking, it prevents the open/closed events (which also avoids the Dart server doing some analysis and producing some notifications we normally want for open files). Thanks!

Out of interest, do you have some approx target for when this client/next LSP will ship? (I'm not in any rush, it's just useful for my planning).

**Comment by dbaeumer** [comment]
After summer vacation :-) (August or September)

**Comment by DanTup** [comment]
Great, thanks! :-)

**Comment by DanTup** [comment]
@dbaeumer no problem, thanks for the update!

**Comment by DanTup** [comment]
@dbaeumer I saw a new `next` version of the client was published last week - does this mean a release is close or are there other things you're planning to do first? There are no open issues in the Next milestone here, but [two in the LSP one](https://github.com/microsoft/vscode-languageserver-node/milestone/67) - however they both just look like spec clarifications and not new work.

FWIW I did some testing of both this change and the `CompletionItem.defaults.data` merging with the latest next against the Dart server, and both are working great 🙂

Thanks!
